src/devices/src/virtio/console/inorder_handler.rs

```diff
@@ -0,0 +1,112 @@
+use std::io::Write;
+use std::result;
+use virtio_console::console;
+use virtio_queue::{Queue, QueueStateOwnedT, QueueStateT};
+use vm_memory::GuestAddressSpace;
+
+#[derive(Debug)]
+pub enum Error {
+    GuestMemory(vm_memory::GuestMemoryError),
+    Queue(virtio_queue::Error),
+    Console(console::Error),
+}
+
+impl From<vm_memory::GuestMemoryError> for Error {
+    fn from(e: vm_memory::GuestMemoryError) -> Self {
+        Error::GuestMemory(e)
+    }
+}
+
+impl From<virtio_queue::Error> for Error {
+    fn from(e: virtio_queue::Error) -> Self {
+        Error::Queue(e)
+    }
+}
+
+impl From<console::Error> for Error {
+    fn from(e: console::Error) -> Self {
+        Error::Console(e)
+    }
+}
+
+pub struct InOrderQueueHandler<M: GuestAddressSpace, S: SignalUsedQueue, T: Write> {
+    pub driver_notify: S,
+    pub transmitq: Queue<M>,
+    pub receiveq: Queue<M>,
+    pub console: console::Console<T>,
+}
+
+impl<M, S, T> InOrderQueueHandler<M, S, T>
+where
+    M: GuestAddressSpace,
+    S: SignalUsedQueue,
+    T: Write,
+{
+    pub fn process_transmitq(&mut self) -> result::Result<(), Error> {
+        // To see why this is done in a loop, please look at the `Queue::enable_notification`
+        // comments in `virtio_queue`.
+        loop {
+            self.transmitq.disable_notification()?;
+
+            while let Some(mut chain) = self
+                .transmitq
+                .state
+                .pop_descriptor_chain(self.transmitq.mem.memory())
+            {
+                self.console.process_transmitq_chain(&mut chain)?;
+
+                self.transmitq.add_used(chain.head_index(), 0)?;
+            }
+            if !self.transmitq.enable_notification()? {
+                break;
+            }
+        }
+        if self.transmitq.needs_notification()? {
+            self.driver_notify.signal_used_queue(1);
+        }
+
+        Ok(())
+    }
+
+    pub fn process_receiveq(&mut self) -> result::Result<(), Error> {
+        // To see why this is done in a loop, please look at the `Queue::enable_notification`
+        // comments in `virtio_queue`.
+        let mut notify = false;
+
+        loop {
+            self.receiveq.disable_notification()?;
+
+            while let Some(mut chain) = self
+                .receiveq
+                .state
+                .pop_descriptor_chain(self.receiveq.mem.memory())
+            {
+                let used_len = match self.console.process_receiveq_chain(&mut chain) {
+                    Ok(used_len) => used_len,
+                    Err(e) => {
+                        self.receiveq.state.go_to_previous_position();
+                        return Err(Error::Console(e));
+                    }
+                };
+                if used_len == 0 {
+                    self.receiveq.state.go_to_previous_position();
+                    break;
+                }
+                self.receiveq
+                    .add_used(chain.head_index(), used_len as u32)?;
```

Comment:
I'm not sure about the conversion here from usize to u32. Can we actually update `process_receiveq_chain` to return an u32 instead? From a quick look, it seemed to me that we can.